### PR TITLE
Propagating changes from v015 

### PR DESCRIPTION
**Pull request by sourabhdalvi**

### CODE REVIEW

File: src/core/variables.jl [hunk 1]
```diff
@@ -25,6 +25,9 @@ get_component_type(::VariableKey{T, U}) where {T <: VariableType, U <: PSY.Compo
 """Struct to dispatch the creation of Active Power Variables"""
 struct ActivePowerVariable <: VariableType end
 
+"""Struct to dispatch the creation of Active Power Variables above minimum power"""
```
Comment (jd-lara) [inline src/core/variables.jl:28]: Mention in the docstring the formulations that use this.

File: src/devices_models/devices/common/add_dual_variable.jl [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    constriant_type::Type{<:ConstraintType},
+    sys::U,
+    formulation,
+) where {U <: PSY.System}
+    time_steps = get_time_steps(container)
+    add_dual_container!(container, constriant_type, U, time_steps)
+    return
+end
+
+function calculate_dual_variable_value!(
```
Comment (jd-lara) [inline src/devices_models/devices/common/add_dual_variable.jl:79]: make this an internal method since it dispatches on the key. 
```suggestion
function _calculate_dual_variable_value!(
```

File: src/devices_models/devices/thermal_generation.jl [hunk 3]
```diff
@@ -517,6 +563,28 @@ function calculate_aux_variable_value!(
 
     return
 end
+
+function calculate_aux_variable_value!(
+    optimization_container::OptimizationContainer,
+    key::AuxVarKey{PowerOutput, T},
```
Comment (jd-lara) [inline src/devices_models/devices/thermal_generation.jl:569]: we don't need the key here. Please use PowerOutput and T separately.

File: src/devices_models/devices/thermal_generation.jl [hunk 4]
```diff
@@ -517,6 +563,28 @@ function calculate_aux_variable_value!(
...
+    system::PSY.System,
+) where {T <: PSY.ThermalGen}
+    devices = PSY.get_components(T, system)
+    time_steps = model_time_steps(optimization_container)
+    on_var_results = get_variable(optimization_container, OnVariable(), T)
+    p_var_results = get_variable(optimization_container, PowerAboveMinimumVariable(), T)
+    aux_var_container = get_aux_variables(optimization_container)[key]
```
Comment (jd-lara) [inline src/devices_models/devices/thermal_generation.jl:576]: use the appropriate getter function with the var type and T

File: src/devices_models/devices/thermal_generation.jl [hunk 5]
```diff
@@ -517,6 +563,28 @@ function calculate_aux_variable_value!(
...
+
+    for d in devices, t in time_steps
+        name = PSY.get_name(d)
+        active_limits = PSY.get_active_power_limits(d)
```
Comment (jd-lara) [inline src/devices_models/devices/thermal_generation.jl:580]:
```suggestion
        min = PSY.get_active_power_limits(d).min
```

File: src/devices_models/devices/thermal_generation.jl [hunk 6]
```diff
@@ -517,6 +563,28 @@ function calculate_aux_variable_value!(
...
+        aux_var_container[name, t] =
+            JuMP.value(on_var_results[name, t]) * active_limits.min +
```
Comment (jd-lara) [inline src/devices_models/devices/thermal_generation.jl:582]:
```suggestion
            JuMP.value(on_var_results[name, t]) * min +
```

File: src/network_models/network_constructor.jl [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
     sys::PSY.System,
-    ::Type{CopperPlatePowerModel},
+    model::NetworkModel{CopperPlatePowerModel},
     template::ProblemTemplate,
 )
     buses = PSY.get_components(PSY.Bus, sys)
     bus_count = length(buses)
 
-    if get_balance_slack_variables(container.settings)
+    if get_use_slacks(model)
```
Comment (jd-lara) [inline src/network_models/network_constructor.jl:10]:
```suggestion
    get_use_slacks(model) && add_slacks!(container, CopperPlatePowerModel)
```

File: src/core/device_model.jl [hunk 8]
```diff
@@ -32,15 +32,15 @@ thermal_gens = DeviceModel(ThermalStandard, ThermalBasicUnitCommitment),
 mutable struct DeviceModel{D <: PSY.Device, B <: AbstractDeviceFormulation}
     feedforward::Union{Nothing, AbstractAffectFeedForward}
     use_slacks::Bool
-    duals::Vector{<:ConstraintType}
+    duals::Vector
```
Comment (daniel-thom) [inline src/core/device_model.jl:35]: Just curious, what is the reason for this change? Are we no longer bounding the types?

File: src/core/optimization_container.jl [hunk 9]
```diff
@@ -409,6 +416,28 @@ function get_aux_variable_keys(container::OptimizationContainer)
     return collect(keys(container.aux_variables))
 end
 
+##################################### DualVariable Container ################################
+function add_dual_container!(
```
Comment (jd-lara) [inline src/core/optimization_container.jl:420]: Wherever this is called we need to check if the container is compatible i.e., check if its MILP

File: src/core/device_model.jl (hunk 8 above)
Comment (jd-lara) [reply src/core/device_model.jl:35]: This is actually a `Vector{DataType}`.

**Review by jd-lara — APPROVED** [review]
There are some minor changes requested. You can merge after those are addressed.

File: src/devices_models/devices/thermal_generation.jl (hunk 3 above)
Comment (sourabhdalvi) [reply src/devices_models/devices/thermal_generation.jl:569]: we do use the key to get the aux container `aux_var_container = get_aux_variables(container)[key]`. This is just following the same template as TimeDurationOff/On

File: src/devices_models/devices/thermal_generation.jl (hunk 4 above)
Comment (sourabhdalvi) [reply src/devices_models/devices/thermal_generation.jl:576]: I'm not sure what you mean, there are only two `get_variable` methods one uses VariableKey and the other is the one I use.

File: src/devices_models/devices/thermal_generation.jl (hunk 3 above)
Comment (jd-lara) [reply src/devices_models/devices/thermal_generation.jl:569]: we then need a method similar to: 

```Julia
function get_variable(
    container::OptimizationContainer,
    ::T,
    ::Type{U},
    meta::String = CONTAINER_KEY_EMPTY_META,
) where {T <: VariableType, U <: PSY.Component}
    return get_variable(container, VariableKey(T, U, meta))
end
```

for get_aux_variable

File: src/core/device_model.jl (hunk 8 above)
Comment (daniel-thom) [reply src/core/device_model.jl:35]: So is it supposed to be defined as `Vector{DataType}`?

Comment (jd-lara) [reply src/core/device_model.jl:35]: yes, @sourabhdalvi we can make this `Vector{DataType}` since it stores types of constraints.

**Review by daniel-thom — APPROVED** [review]

File: test/test_operations_problem.jl [hunk 10]
```diff
@@ -151,18 +141,19 @@ end
     parameters = [true, false]
     ptdf = PTDF(sys)
     # These are the duals of interest for the test
-    dual_constraint =
-        [[:nodal_balance_active__Bus], [:CopperPlateBalance, :network_flow__Line]]
+    dual_constraint = [[NodalBalanceActiveConstraint], [CopperPlateBalanceConstraint]]
```
Comment (daniel-thom) [inline test/test_operations_problem.jl:144]: @sourabhdalvi @jd-lara I'm now working with this code. Should this interface take a ConstraintKey instead of ConstraintType (or tuple that can be used to create a ConstraintKey)? Before this PR I had written some code like this:
```
dual_constraint = [[(NodalBalanceActiveConstraint, PSY.Bus)], [(CopperPlateBalanceConstraint, PSY.System]]
```

File: test/test_operations_problem.jl [hunk 11]
```diff
@@ -171,52 +162,45 @@ end
...
 
     container = PSI.get_optimization_container(model)
-    constraints = PSI.get_constraints(container)[PSI.ConstraintKey(
-        CopperPlateBalanceConstraint,
-        PSY.System,
-    )]
-    dual_results = read_duals(container, [:CopperPlateBalance])
+    constraint_key = PSI.ConstraintKey(CopperPlateBalanceConstraint, PSY.System)
+    constraints = PSI.get_constraints(container)[constraint_key]
+    dual_results = read_duals(container)[:CopperPlateBalanceConstraint_System]
```
Comment (daniel-thom) [inline test/test_operations_problem.jl:185]: @sourabhdalvi @jd-lara Similar to above, I was planning to have something like
```
dual_results = read_duals(container)[(CopperPlateBalanceConstraint, PSY.System)]
```

File: test/test_operations_problem.jl (hunk 10 above)
Comment (jd-lara) [reply test/test_operations_problem.jl:144]: Since the dual constraint type gets passed with the model we can build the key at the construction time. The only exception is the case of the network where the device portion can be a system or a bus depending on the constraint type we want de dual from. 

I think we can still pass the constraint type and fill the remaining portion after since the system applies only in the case of the CopperPlateBalanceConstraint and Bus in any other constraint type.

File: test/test_operations_problem.jl (hunk 11 above)
Comment (jd-lara) [reply test/test_operations_problem.jl:185]: forgetting the results I think this is fine, albeit inconsistent with the other interface but the dual interface will be for programmatic querying of the duals and it isn't too bad.

Comment (daniel-thom) [reply test/test_operations_problem.jl:185]: Shouldn't the interfaces in the user code be symmetric? Otherwise, how will they know what to pass to `read_duals`?

To be clear I mean
```
NetworkModel(network; PTDF = ptdf, duals = (CopperPlateBalanceConstraint, PSY.System))
read_duals(container, (CopperPlateBalanceConstraint, PSY.System))
```

Comment (jd-lara) [reply test/test_operations_problem.jl:185]: I get what you mean 100% and agree that there is inconsistency. I am not sure what is worse that's all because in the case of the device and service model we don't need to pass the device type for the duals.
